### PR TITLE
Automatically create raft groups for incoming messages.

### PR DESCRIPTION
**Pull request by bdarnell**
Enable the receiving of snapshots in multiraft.

Reinstate multiraft.TestMembershipChange.

@spencerkimball @cockroachdb/developers

### CODE REVIEW

**Comment by spencerkimball** [comment]
Looks like same error you were trying to fix with the util.Stopper on tests. Is that change already in?

File: multiraft/multiraft.go [hunk 1]
```diff
@@ -409,16 +409,15 @@ func (s *state) start() {
...
-			_, nodeOK := s.nodes[NodeID(req.Message.From)]
-			groupOK = groupOK || req.Message.Type == raftpb.MsgHeartbeat
-			if !groupOK || !nodeOK || NodeID(req.Message.To) != s.nodeID {
-				log.V(4).Infof("node %v: discarding ill-addressed message %s", s.nodeID, req.GroupID,
-					raft.DescribeMessage(req.Message, s.EntryFormatter))
-				break
-			}
 			log.V(5).Infof("node %v: group %v got message %s", s.nodeID, req.GroupID,
 				raft.DescribeMessage(req.Message, s.EntryFormatter))
+			if _, ok := s.groups[req.GroupID]; !ok {
```
Comment (spencerkimball) [inline multiraft/multiraft.go]: Do you not care anymore about not knowing about the node which the message was sent from?

**Comment by spencerkimball** [comment]
LGTM

File: multiraft/multiraft.go (hunk 1 above)
Comment (bdarnell) [inline multiraft/multiraft.go]: No; it is possible for our first contact with a new node to be when that node adds us to a new group.

**Comment by bdarnell** [comment]
Yes, the failure here is the one that prompted #312, although per the comments in that thread I no longer think that's the right fix. I'll take a closer look.

**Comment by bdarnell** [comment]
The test failures in this branch have been fixed with #318, so I'm merging it now.

**Pull request merged**